src/Lecture2.hs

```diff
@@ -164,7 +186,46 @@ data Knight = Knight
     , knightEndurance :: Int
     }
 
-dragonFight = error "TODO"
+data Chest a = Chest
+  {
+    chestGold :: Int
+    ,chestTreasure :: a
+  }
+data Dragon a = Dragon
+  {
+    dragonColor :: String ,
+    dragonHealth :: Int,
+    dragonFirePower :: Int,
+    dragonExpPoints :: Int,
+    dragonChest :: Chest a
+  }
+
+data DragonType = RedDragon | BlackDragon | GreenDragon
+
+type Armor = Int 
+type HealthPotion = Int 
+
+dragon :: DragonType -> Dragon (a, b)
+dragon d = case d of
+  RedDragon -> Dragon "Red" 100 500 15 (Chest 100 Armor)
```

Comment:
You have few errors here:

1. You specified the type of treasure (via the `Dragon` type variable that will be propagated to treasure) as a pair of polymorphic variables `a` and `b` while you construct the chest with the value `Armor` (which is not pair)
2. There's no constructor `Armor` for any of the types. `Armor` is the name of the type (currently an alias to `Int`).

So, a correct way to write this code so it'll be compiled is the following:

```haskell
dragon :: DragonType -> Dragon Armor
dragon d = case d of
  RedDragon -> Dragon "Red" 100 500 15 (Chest 100 249)
```

We're specifying that the dragon is parametrised by the `Armor` type (so the `Chest` inside will store `Armor`) and the actual armor value is `249`.